### PR TITLE
Place Token on Board

### PR DESCRIPTION
**Pull request by DeKaylaPhillips**
# Place Token on Board

## Description

Main branch currently supplies the ability to display a standard Tic-Tac-Toe board with modifiable cell values. The main branch lacks the necessary implementation needed to update the board's display. The 'feature-place-token' branch adds this functionality, allowing for a valid token object at an unoccupied position to be placed on the board. 

## Type of Change

- New Feature (non-breaking change which adds functionality)

## Tests

**Test Suite: `board.test.js`**

- Tests that specific rows and columns are able to be identified retrieved based on a position in the board between 1-9.

- Tests that a token can be placed on the board at some position when the token is a string and position is unoccupied and between 1-9.

**Test Suite: `token.test.js`**

- Tests the constructor of the Token class will accept and assign the player's token value to a validated token value.

- Tests the constructor of the Token class will assign the player's token to a null value when a token is invalid.

- Tests the getToken() method will return a string token with a length of 1.

### CODE REVIEW

File: test/Board/board.test.js [hunk 1]
```diff
@@ -1,54 +1,53 @@
...
+            expect(cellStructure).toBe(board.cells);
+        });    
+    });
 
-        test('will print a standard, user-friendly 3x3 Tic-Tac-Toe board', () => {
-            const printedBoard = board.toString()
-            const rowOutput = `   |    |  `
-            expect(printedBoard).toBe(`${rowOutput}\n${board.rowSeparator}\n${rowOutput}\n${board.rowSeparator}\n${rowOutput}`)
+    describe('getRow()', () => {
+        test('will return the correlated row number for a valid position', () => {
```
Comment (njdennis-cars) [inline test/Board/board.test.js:33]: @DeKaylaPhillips Can you explain what this test is doing?

Comment (Naomi-Dennis) [inline test/Board/board.test.js:33]: @DeKaylaPhillips Can you explain what this function is doing? Why does `board.getRow(#)` return a number?

File: test/Board/board.test.js [hunk 2]
```diff
@@ -1,54 +1,53 @@
...
+            const row = board.getRow(4);
+            expect(row).toBe(1);
+        });
+    });
+ 
+    describe('getColumn()', () => {
+        test('will return the correlated column number for a valid position', () => {
```
Comment (Naomi-Dennis) [inline test/Board/board.test.js:40]: @DeKaylaPhillips Similar question to the one for `getRow()`. Can you explain what this function is doing? Why does `board.Column(#)` return a number?

File: test/Board/board.test.js [hunk 3]
```diff
@@ -1,54 +1,53 @@
...
-            board.cells[0][0] = 'X' 
-          
-            const newBoardDisplay = board.assembleBoard()
-
-            expect(initialBoardDisplay).not.toBe(newBoardDisplay) 
+    describe('placeToken()', () => {
+        test('will place a valid token in the board when a valid position is accepted', () => {
+            board.placeToken(new Token('X').getToken(), 2);
+            const cells = board.getCells();
+            expect(cells).toStrictEqual([['', 'X', ''],['', '', ''],['', '', '']]);
```
Comment (Naomi-Dennis) [inline test/Board/board.test.js:50]: 🎉 This test is great!

File: test/Board/board.test.js (hunk 2 above)
Comment (DeKaylaPhillips) [reply test/Board/board.test.js:40]: Sure thing!

Similarly to the `getRow()` method, the `getColumn` method returns the results of the private #findRow() method which is used to retrieve the correct index number of a row in the cells array based on the position number (1-9) passed to it as an argument.

#findIndex looks through an object containing the maximum position number that would represent the final column in a row and retrieves the index number of one of the three inner array elements [[ ], [ ], [ ]] (row.rowIdx) if the given position number is either less than (which would be one of the first cells/columns*) or equal to the last cell/column in the array.

Positions 1, 2, 3 ---> [[1, 2, 3] --> Index 0 returned
Positions 4, 5, 6 ---> [4, 5, 6], --> Index 1 returned
Positions 7, 8, 9 ---> [7, 8, 9]] --> Index 2 returned

The number returned from this method is used as the first index number of the nested cells array (...this.cells[-row#-]...) when accessing a particular 'row' array element in the nested cells array.

Comment (DeKaylaPhillips) [reply test/Board/board.test.js:40]: Sure, thing!

Similar to the `getRow()` method, the `getColumn()` method returns the results of the private #findColumn() method which is used to retrieve the correct index number of a column where a cell would live in the cells array based on the position number (1-9) passed to it as an argument.

`#findColumn` looks through an object containing row index numbers and a `column` key whose value calculates the index of a cell element at a certain given position by subtracting the given position number passed as an argument from the minimum position number that could lie within a row. 

Row 0 ---> [[1, 2, 3] --> Index 0, 1, or 2 returned
Row 1 ---> [4, 5, 6], --> Index 0, 1, or 2 returned
Row 2 ---> [7, 8, 9]] --> Index 0, 1, or 2 returned

The column index number returned from this method is used as the second index number of the nested cells array (... this.cells[-row#-][-col#-] ...) when accessing a particular cell at some column within a row array element in the nested cells array.

File: lib/Board/board.js [hunk 4]
```diff
@@ -6,31 +6,77 @@ export class Board {
...
+    // public getCells(): Array 
     getCells() {
         return this.cells;
     };
 
-    /* toString():String */
-    toString() {
-        return this.assembleBoard();
+    // public getRow(): Number
+    getRow(position) {
```
Comment (Naomi-Dennis) [inline lib/Board/board.js:20]: The type for `getRow()` should take an argument! What's the argument data type?

File: lib/Board/board.js [hunk 5]
```diff
@@ -6,31 +6,77 @@ export class Board {
...
+        return this.#findRow(position);
+    };
+
+    // public getColumn(): Number
+    getColumn(position) {
```
Comment (Naomi-Dennis) [inline lib/Board/board.js:25]: The type for `getColumn()` should take an argument! What's the argument's data type?

**Review by Naomi-Dennis — APPROVED** [review]

**Pull request merged**